crates/ruma-events/src/room/message.rs

```diff
@@ -188,33 +318,33 @@ impl RoomMessageEventContent {
             (*body, *formatted_body) = reply::plain_and_formatted_reply_body(
                 body.as_str(),
                 (!formatted_body.is_empty()).then_some(formatted_body.as_str()),
-                original_message,
+                original_event,
             );
         }
 
-        let relates_to = if let Some(Relation::Thread(Thread { event_id, .. })) = original_message
-            .content
-            .relates_to
-            .as_ref()
-            .filter(|_| forward_thread == ForwardThread::Yes)
-        {
-            Relation::Thread(Thread::plain(event_id.clone(), original_message.event_id.clone()))
+        self
+    }
+
+    fn make_reply_tweaks(
+        mut self,
+        original_event_id: OwnedEventId,
+        original_thread_id: Option<OwnedEventId>,
+        original_user_mentions: Option<BTreeSet<OwnedUserId>>,
+        original_sender: Option<&UserId>,
+    ) -> Self {
+        let relates_to = if let Some(event_id) = original_thread_id {
+            Relation::Thread(Thread::plain(event_id.to_owned(), original_event_id.to_owned()))
         } else {
-            Relation::Reply {
-                in_reply_to: InReplyTo { event_id: original_message.event_id.clone() },
-            }
+            Relation::Reply { in_reply_to: InReplyTo { event_id: original_event_id.to_owned() } }
         };
         self.relates_to = Some(relates_to);
 
-        if add_mentions == AddMentions::Yes {
-            // Copy the mentioned users.
-            let mut user_ids = match &original_message.content.mentions {
-                Some(m) => m.user_ids.clone(),
-                None => Default::default(),
-            };
-            // Add the sender.
-            user_ids.insert(original_message.sender.clone());
-            self.mentions = Some(Mentions { user_ids, ..Default::default() });
+        if let Some(sender) = original_sender {
+            if let Some(mut user_ids) = original_user_mentions {
```

Comment:
```suggestion
        if let (Some(sender), Some(mut user_ids)) = (original_sender, original_user_mentions) {
```

though an `if let` chain would be nicer in a way (wondering when it will become stable...)